src/react/Reconciler.js

```diff
@@ -0,0 +1,293 @@
+import { GetValue, Get } from "../methods/index.js";
+import invariant from "../invariant.js";
+import { flowAnnotationToObjectTypeTemplate } from "../flow/utils.js";
+import { computeBinary } from "../evaluators/BinaryExpression.js";
+import * as t from "babel-types";
+import type { BabelNodeIdentifier } from "babel-types";
+import { createAbstractObject } from "../flow/abstractObjectFactories.js";
+import { CompilerDiagnostic, FatalError } from "../errors.js";
+
+// Branch status is used for when Prepack returns an abstract value from a render
+// that results in a conditional path occuring. This can be problematic for reconcilation
+// as the reconciler then needs to understand if this is the start of a new branch, or if
+// it's actually deep into an existing branch. If it's a new branch, we need to apply
+// keys to the root JSX element so that it keeps it identity (because we're folding trees).
+// Furthermore, we also need to bail-out of folding class components where they have lifecycle
+// events, as we can't merge lifecycles of mutliple trees when branched reliably
+const BranchStatus = {
+  NO_BRANCH: "NO_BRANCH",
+  NEW_BRANCH: "NEW_BRANCH",
+  BRANCH: "BRANCH",
+};
+
+type BranchStatusEnum = $Keys<typeof BranchStatus>;
+
+// ExpectedBailOut is like an error, that gets thrown during the reconcilation phase
+// allowing the reconcilation to continue on other branches of the tree, the message
+// given to ExpectedBailOut will be assigned to the value.$BailOut property and serialized
+// as a comment in the output source to give the user hints as to what they need to do
+// to fix the bail-out case
+class ExpectedBailOut {
+  message: string;
+  constructor(message: string) {
+    this.message = message;
+  }
+}
+
+class Reconciler {
+  constructor(
+    realm: Realm,
+    moduleTracer: ModuleTracer,
+    statistics: ReactStatistics,
+    reactSerializerState: ReactSerializerState
+  ) {
+    this.realm = realm;
+    this.moduleTracer = moduleTracer;
+    this.statistics = statistics;
+    this.reactSerializerState = reactSerializerState;
+  }
+
+  realm: Realm;
+  moduleTracer: ModuleTracer;
+  statistics: ReactStatistics;
+  reactSerializerState: ReactSerializerState;
+
+  render(componentType: ECMAScriptSourceFunctionValue): Effects {
+    let propTypes = null;
+    let propsName = null;
+    let contextTypes = null;
+    let contextName = null;
+    // we take the first "props" paramater from "function MyComponent (props, context)" and look at its name
+    // if its not an Identifier, we leave propsName null so it doesn't get used to create the object
+    if (componentType.$FormalParameters.length > 0) {
+      if (t.isIdentifier(componentType.$FormalParameters[0])) {
+        propsName = ((componentType.$FormalParameters[0]: any): BabelNodeIdentifier).name;
+      }
+      // we expect that if there's a props argument, it should always have Flow annotations
+      invariant(
```

Comment:
Is that because Flow-types are turned on when we get here, and even if the user omits it, some Babel transform insert an `any` type?